server/coordinator/shard_picker.go

```diff
@@ -19,83 +18,41 @@ type ShardPicker interface {
 	PickShards(ctx context.Context, snapshot metadata.Snapshot, expectShardNum int) ([]storage.ShardNode, error)
 }
 
-// RandomBalancedShardPicker randomly pick up shards that are not on the same node in the current cluster.
-type RandomBalancedShardPicker struct{}
+// LeastTableShardPicker selects shards based on the number of tables on the current shard,
+// and always selects the shard with the smallest number of current tables.
+type leastTableShardPicker struct{}
 
-func NewRandomBalancedShardPicker() ShardPicker {
-	return &RandomBalancedShardPicker{}
+func NewLeastTableShardPicker() ShardPicker {
+	return &leastTableShardPicker{}
 }
 
-// PickShards will pick a specified number of shards as expectShardNum.
-func (p *RandomBalancedShardPicker) PickShards(_ context.Context, snapshot metadata.Snapshot, expectShardNum int) ([]storage.ShardNode, error) {
+func (l leastTableShardPicker) PickShards(_ context.Context, snapshot metadata.Snapshot, expectShardNum int) ([]storage.ShardNode, error) {
+	shardNodeMapping := make(map[storage.ShardID]storage.ShardNode, len(snapshot.Topology.ShardViewsMapping))
+	for _, shardNode := range snapshot.Topology.ClusterView.ShardNodes {
+		shardNodeMapping[shardNode.ID] = shardNode
+	}
+
 	if len(snapshot.Topology.ClusterView.ShardNodes) == 0 {
 		return nil, errors.WithMessage(ErrNodeNumberNotEnough, "no shard is assigned")
 	}
 
-	shardNodes := snapshot.Topology.ClusterView.ShardNodes
-
-	nodeShardsMapping := make(map[string][]storage.ShardNode, 0)
-	for _, shardNode := range shardNodes {
-		_, exists := nodeShardsMapping[shardNode.NodeName]
-		if !exists {
-			nodeShardsMapping[shardNode.NodeName] = []storage.ShardNode{}
-		}
-		nodeShardsMapping[shardNode.NodeName] = append(nodeShardsMapping[shardNode.NodeName], shardNode)
+	sortedShardsByTableCount := make([]storage.ShardID, 0, len(snapshot.Topology.ShardViewsMapping))
+	for shardID := range snapshot.Topology.ShardViewsMapping {
```

Comment:
Shall we only collect the shards which has been allocated to a node? That is to say, we should loop over the shards of the `shardNodeMapping`. Otherwise, L54 may not find the correct shard node.